### PR TITLE
Feature/uar 612 handling a change to no change

### PR DESCRIPTION
**Pull request by olaych**
### JIRA link

https://companieshouse.atlassian.net/browse/UAR-612

### Change description
Added reset data for changed and no change scenarios

### Work checklist

- [x] Tests added where applicable
- [x] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

**Review by bbell-companieshouse — APPROVED** [review]

File: src/controllers/update/overseas.entity.change.controller.ts [hunk 1]
```diff
@@ -49,3 +52,33 @@ export const post = async (req: Request, resp: Response, next: NextFunction) =>
...
+  if (appData){
+    appData.who_is_registering = undefined;
+    appData.overseas_entity_due_diligence = undefined;
+    appData.beneficial_owners_individual = undefined;
+    appData.beneficial_owners_corporate = undefined;
+    appData.beneficial_owners_government_or_public_authority = undefined;
+    appData.managing_officers_corporate = undefined;
+    appData.managing_officers_individual = undefined;
+    appData.beneficial_owners_statement = undefined;
+    appData.payment = undefined;
```
Comment (AndreKramer7654) [inline src/controllers/update/overseas.entity.change.controller.ts:66]: The OE entity could be changed also so do:
const companyProfile = await getCompanyProfile(req, appData.entityNumber);
await addOeToApplicationData(req, appData, appData.entityNumber, companyProfile);

Comment (AndreKramer7654) [reply src/controllers/update/overseas.entity.change.controller.ts:66]: appData.due_diligence = undefined
appData.overseas_entity_due_diligence = undefined

// appData.presenter is entered but is ok
// maybe so we don't forget in future
appData.trusts = undefined

File: src/controllers/update/overseas.entity.change.controller.ts [hunk 2]
```diff
@@ -49,3 +52,33 @@ export const post = async (req: Request, resp: Response, next: NextFunction) =>
...
+    appData.beneficial_owners_statement = undefined;
+    appData.update.registrable_beneficial_owner = undefined;
+    appData.payment = undefined;
+  }
+  return appData;
+};
+
+const existingBoMoForNoChange = async (req: Request, appData: ApplicationData) => {
+  await retrieveBeneficialOwners(req, appData);
+  await retrieveManagingOfficers(req, appData);
```
Comment (AndreKramer7654) [inline src/controllers/update/overseas.entity.change.controller.ts:83]: appData.update.bo_mo_data_fetched = true;

File: src/controllers/update/overseas.entity.change.controller.ts (hunk 1 above)
Comment (AndreKramer7654) [reply src/controllers/update/overseas.entity.change.controller.ts:66]: appData.update.bo_mo_data_fetched = false;

Comment (olaych) [reply src/controllers/update/overseas.entity.change.controller.ts:66]: But won't the OE change be handled before no_change journey. The reset will take place on the OE query page I suppose? I would think only the data from no change page is required to reset?

Comment (AndreKramer7654) [reply src/controllers/update/overseas.entity.change.controller.ts:66]: // appData.update.date_of_creation, appData.entity_name and appData.entity_number won't be affected.
// appData.update.no_change will be toggled anyway
// appData.filing_date wil not be affected either

appData.update.registrable_benefitial_owner needs to be reset to undefined

Comment (AndreKramer7654) [reply src/controllers/update/overseas.entity.change.controller.ts:66]: I did not mean OE change as in switching to another OE altogether. I meant the user changing something in the OE data - like the email address or other change links on the "OE revirew" page. So re-load the same OE.

File: src/controllers/update/overseas.entity.change.controller.ts [hunk 3]
```diff
@@ -49,3 +52,40 @@ export const post = async (req: Request, resp: Response, next: NextFunction) =>
     next(errors);
   }
 };
+
+export const resetChangeData = async (req: Request, appData: ApplicationData) => {
+  if (appData && appData.update){
```
Comment (AndreKramer7654) [inline src/controllers/update/overseas.entity.change.controller.ts:57]: I'm not sure appData.update is defined when we get to here. Safer to do the appData and appData.update in separate if statements?

File: src/controllers/update/overseas.entity.change.controller.ts [hunk 4]
```diff
@@ -49,3 +52,40 @@ export const post = async (req: Request, resp: Response, next: NextFunction) =>
...
+    appData.trusts = undefined;
+    appData.update.registrable_beneficial_owner = undefined;
+    appData.update.bo_mo_data_fetched = false;
+    await existingBoMoForNoChange(req, appData);
+  }
+  return appData;
+};
+
+export const resetNoChangeData = (appData: ApplicationData) => {
+  if (appData.update){
```
Comment (AndreKramer7654) [inline src/controllers/update/overseas.entity.change.controller.ts:77]: Again, I'm not sure update will be defined always. Maybe take all but appData.update out of if statement?

**Review by AndreKramer7654 — APPROVED** [review]

File: src/controllers/update/overseas.entity.change.controller.ts [hunk 5]
```diff
@@ -50,3 +55,44 @@ export const post = async (req: Request, resp: Response, next: NextFunction) =>
...
+    appData.due_diligence = undefined;
+    appData.overseas_entity_due_diligence = undefined;
+    appData.beneficial_owners_individual = undefined;
+    appData.beneficial_owners_corporate = undefined;
+    appData.beneficial_owners_government_or_public_authority = undefined;
+    appData.managing_officers_corporate = undefined;
+    appData.managing_officers_individual = undefined;
+    appData.beneficial_owners_statement = undefined;
+    appData.payment = undefined;
+    appData.trusts = undefined;
```
Comment (AndreKramer7654) [inline src/controllers/update/overseas.entity.change.controller.ts:71]: appData.entity = undefined; 
just to make sure?

**Pull request merged**